### PR TITLE
ENG-11718: remove conversions from jar to bytes or bytes to jar when …

### PR DESCRIPTION
**Pull request by xinjiacs**
…they are unnecessary

There are so redundant conversion between jar to bytes. This PR tries to minimize the conversions which take time.

### CODE REVIEW

File: src/frontend/org/voltdb/utils/InMemoryJarfile.java [hunk 1]
```diff
@@ -443,4 +447,13 @@ public Object clone() {
...
         throw new UnsupportedOperationException();
     }
+
+    public InMemoryJarfile deepCopy () {
+        InMemoryJarfile copy = new InMemoryJarfile();
+        for (Entry<String, byte[]> e: this.entrySet()) {
+            copy.put(e.getKey(), e.getValue());
+        }
+        copy.m_loader.initFrom(m_loader);
+        return copy;
```
Comment (zeemanhe) [inline src/frontend/org/voltdb/utils/InMemoryJarfile.java]: Is this more like shallow copy?

File: src/frontend/org/voltdb/compiler/AsyncCompilerAgentHelper.java [hunk 2]
```diff
@@ -281,16 +265,14 @@ else if (work.invocationName.startsWith("@AdHoc")) {
 
         VoltCompiler compiler = new VoltCompiler();
         compiler.compileInMemoryJarfileWithNewDDL(jarfile, newDDL, oldCatalog);
-        return jarfile.getFullJarBytes();
+        return jarfile;
     }
 
-    private byte[] modifyCatalogClasses(byte[] oldCatalogBytes, String deletePatterns,
-            byte[] newClassBytes) throws IOException
+    private InMemoryJarfile modifyCatalogClasses(InMemoryJarfile jarfile, String deletePatterns,
```
Comment (zeemanhe) [inline src/frontend/org/voltdb/compiler/AsyncCompilerAgentHelper.java]: No return seems to be needed for  addDDLToCatalog()m and modifyCatalogClasses. change them to void.

File: src/frontend/org/voltdb/compiler/AsyncCompilerAgentHelper.java [hunk 3]
```diff
@@ -62,40 +62,29 @@ public CatalogChangeResult prepareApplicationCatalogDiff(CatalogChangeWork work)
...
-                if (newCatalogBytes == null) {
-                    try {
-                        newCatalogBytes = context.getCatalogJarBytes();
-                    }
-                    catch (IOException ioe) {
-                        retval.errorMsg = "Unexpected exception retrieving internal catalog bytes: " +
-                            ioe.getMessage();
-                        return retval;
-                    }
+                // Grab the current catalog bytes if @UAC had a null catalog from deployment-only update
```
Comment (zeemanhe) [inline src/frontend/org/voltdb/compiler/AsyncCompilerAgentHelper.java]: "@UpdateApplicationCatalog".equals(work.invocationName) is safer.

File: src/frontend/org/voltdb/compiler/AsyncCompilerAgentHelper.java [hunk 4]
```diff
@@ -62,40 +62,29 @@ public CatalogChangeResult prepareApplicationCatalogDiff(CatalogChangeWork work)
...
+                if (work.operationBytes == null) {
+                    newCatalogJar = oldJar;
+                } else {
+                    newCatalogJar = CatalogUtil.loadInMemoryJarFile(work.operationBytes);
                 }
                 // If the deploymentString is null, we'll fill it in with current deployment later
                 // Otherwise, deploymentString has the right contents, don't need to touch it
             }
             else if (work.invocationName.equals("@UpdateClasses")) {
```
Comment (zeemanhe) [inline src/frontend/org/voltdb/compiler/AsyncCompilerAgentHelper.java]: "@UpdateClasses".equals(work.invocationName) is safer.

File: src/frontend/org/voltdb/compiler/AsyncCompilerAgentHelper.java [hunk 5]
```diff
@@ -128,8 +115,8 @@ else if (work.invocationName.startsWith("@AdHoc")) {
                     compilerLog.error(retval.errorMsg);
                     return retval;
                 }
-                assert(newCatalogBytes != null);
-                if (newCatalogBytes == null) {
+                assert(newCatalogJar != null);
+                if (newCatalogJar == null) {
```
Comment (zeemanhe) [inline src/frontend/org/voltdb/compiler/AsyncCompilerAgentHelper.java]: Is the assert needed?

File: src/frontend/org/voltdb/utils/InMemoryJarfile.java (hunk 1 above)
Comment (xinjiacs) [inline src/frontend/org/voltdb/utils/InMemoryJarfile.java]: initFrom uses addAll that copies the contents. Why shallow copy?

**Pull request merged**